[Bug] Notification service crash the app if it runs during sync

###### Reproduction Steps

0. Turn on notification
1. Reinstall AnkiDroid
2. Start ankidroid
3. Sync


###### Expected Result

Syncing.
A notification.

Not crashing

###### Actual Result

Crash with the stacktrace:

> 2021-03-18 04:53:05.640 22370-22370/com.ichi2.anki E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.ichi2.anki, PID: 22370
    java.lang.RuntimeException: Unable to start receiver com.ichi2.anki.services.ReminderService: java.lang.NullPointerException: Attempt to invoke virtual method 'int com.ichi2.libanki.DB.queryScalar(java.lang.String, java.lang.Object[])' on a null object reference
        at android.app.ActivityThread.handleReceiver(ActivityThread.java:4207)
        at android.app.ActivityThread.access$1600(ActivityThread.java:273)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2107)
        at android.os.Handler.dispatchMessage(Handler.java:107)
        at android.os.Looper.loop(Looper.java:237)
        at android.app.ActivityThread.main(ActivityThread.java:8107)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:496)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1100)
     Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'int com.ichi2.libanki.DB.queryScalar(java.lang.String, java.lang.Object[])' on a null object reference
        at com.ichi2.libanki.sched.SchedV2._lrnForDeck(SchedV2.java:1478)
        at com.ichi2.libanki.sched.SchedV2.deckDueList(SchedV2.java:552)
        at com.ichi2.libanki.sched.SchedV2.deckDueTree(SchedV2.java:592)
        at com.ichi2.libanki.sched.SchedV2.deckDueTree(SchedV2.java:587)
        at com.ichi2.anki.services.ReminderService.getDeckOptionDue(ReminderService.java:172)
        at com.ichi2.anki.services.ReminderService.onReceive(ReminderService.java:110)
        at android.app.ActivityThread.handleReceiver(ActivityThread.java:4198)
        at android.app.ActivityThread.access$1600(ActivityThread.java:273) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2107) 
        at android.os.Handler.dispatchMessage(Handler.java:107) 
        at android.os.Looper.loop(Looper.java:237) 
        at android.app.ActivityThread.main(ActivityThread.java:8107) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:496) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1100) 

My understanding is that the notification service started running. However, sync closed the database, which means that notification suddenly can't access the database anymore

One way to deal with it would be to try/catch the NullPointerException here and state that notification is not so important that it should crash the program. Or that sync interact with notification, to pause notification until database is reoladed

--- a/AnkiDroid/src/main/java/com/ichi2/anki/CollectionHelper.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/CollectionHelper.java
@@ -400,4 +400,8 @@ public class CollectionHelper {
         }
     }
 
+    @VisibleForTesting(otherwise = VisibleForTesting.NONE)
+    public void setColForTests(Collection col) {
+        this.mCollection = col;
+    }
 }

--- a/AnkiDroid/src/main/java/com/ichi2/anki/services/ReminderService.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/services/ReminderService.java
@@ -169,9 +169,9 @@ public class ReminderService extends BroadcastReceiver {
             return null;
         }
 
-        List<DeckDueTreeNode> dues = col.getSched().deckDueTree();
-        List<DeckDueTreeNode> decks = new ArrayList<>(dues.size());
         try {
+            List<DeckDueTreeNode> dues = col.getSched().deckDueTree();
+            List<DeckDueTreeNode> decks = new ArrayList<>(dues.size());
             // This loop over top level deck only. No notification will ever occur for subdecks.
             for (DeckDueTreeNode node : dues) {
                 JSONObject deck = col.getDecks().get(node.getDid(), false);

--- a/AnkiDroid/src/test/java/com/ichi2/anki/services/ReminderServiceTest.java
+++ b/AnkiDroid/src/test/java/com/ichi2/anki/services/ReminderServiceTest.java
@@ -4,7 +4,9 @@ import android.app.NotificationManager;
 import android.content.Context;
 import android.content.Intent;
 
+import com.ichi2.anki.CollectionHelper;
 import com.ichi2.anki.RobolectricTest;
+import com.ichi2.libanki.Collection;
 
 import org.junit.Test;
 import org.junit.runner.RunWith;
@@ -15,6 +17,10 @@ import androidx.test.ext.junit.runners.AndroidJUnit4;
 import static com.ichi2.anki.services.ReminderService.EXTRA_DECK_ID;
 import static com.ichi2.anki.services.ReminderService.EXTRA_DECK_OPTION_ID;
 import static org.hamcrest.MatcherAssert.assertThat;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
 import static org.robolectric.Shadows.shadowOf;
 
 @RunWith(AndroidJUnit4.class)
@@ -44,6 +50,27 @@ public class ReminderServiceTest extends RobolectricTest {
         buildDefaultDeckReminders();
         // The collection was null so no reminders, but we should get here without exception
         assertThat("No notifications exist", getNotificationManagerShadow().size() == 0);
+    }
+
+
+    /**
+     * #8264: Crash on sync - getSched().getDueTree() failed
+     */
+    @Test
+    public void testDatabaseFailureWhileSyncingDoesNotCrash() {
+        // If getCol() fails, it triggers different exception handling in the service.
+        // The cause was getSched().deckDueTree()
+        Collection baseCol = getCol();
+        Collection mockCol = spy(baseCol);
+        when(mockCol.getSched()).thenThrow(new IllegalStateException("Unit test: simulating database exception"));
+
+        CollectionHelper.getInstance().setColForTests(mockCol);
+
+        buildDefaultDeckReminders();
+
+        // We retry after a database timeout so getSched is called twice
+        //noinspection ResultOfMethodCallIgnored
+        verify(mockCol, times(2)).getSched();
     }
 
 

